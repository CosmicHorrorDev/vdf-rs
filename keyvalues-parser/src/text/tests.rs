use std::{error::Error, fs, path::Path};

use crate::Vdf;

type BoxedResult<T> = Result<T, Box<dyn Error>>;

fn read_asset_file(file_name: &str) -> BoxedResult<String> {
    let val = fs::read_to_string(Path::new("tests").join("assets").join(file_name))?;
    Ok(val)
}

fn snapshot_test_parse_and_render(file_name: &str) -> BoxedResult<()> {
    let vdf_text = read_asset_file(file_name)?;
    let vdf = Vdf::parse(&vdf_text)?;
    insta::assert_ron_snapshot!(vdf);

    let rendered = vdf.to_string();
    insta::assert_snapshot!(rendered);

    Ok(())
}

macro_rules! parse_render_test {
    ($func_name:ident, $file_name: literal) => {
        #[test]
        fn $func_name() -> BoxedResult<()> {
            snapshot_test_parse_and_render($file_name)
        }
    };
}

parse_render_test!(basic, "basic.vdf");
parse_render_test!(app_manifest, "app_manifest.vdf");
parse_render_test!(comments, "comments.vdf");
parse_render_test!(unquoted_strings, "unquoted_strings.vdf");
parse_render_test!(special_charaters, "special_characters.vdf");
parse_render_test!(app_info, "app_info.vdf");
