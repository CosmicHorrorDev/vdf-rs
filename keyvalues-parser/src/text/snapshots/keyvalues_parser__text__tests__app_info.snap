---
source: keyvalues-parser/src/text/tests.rs
expression: vdf

---
Vdf(
  key: "896660",
  value: Obj({
    "common": [
      Obj({
        "ReleaseState": [
          Str("released"),
        ],
        "associations": [
          Obj({}),
        ],
        "clienticon": [
          Str("c55a6b50b170ac6ed56cf90521273c30dccb5f12"),
        ],
        "clienttga": [
          Str("35e067b9efc8d03a9f1cdfb087fac4b970a48daf"),
        ],
        "gameid": [
          Str("896660"),
        ],
        "icon": [
          Str("1aab0586723c8578c7990ced7d443568649d0df2"),
        ],
        "logo": [
          Str("233d73a1c963515ee4a9b59507bc093d85a4e2dc"),
        ],
        "logo_small": [
          Str("233d73a1c963515ee4a9b59507bc093d85a4e2dc_thumb"),
        ],
        "name": [
          Str("Valheim Dedicated Server"),
        ],
        "osarch": [
          Str(""),
        ],
        "oslist": [
          Str("windows,linux"),
        ],
        "parent": [
          Str("892970"),
        ],
        "type": [
          Str("Tool"),
        ],
      }),
    ],
    "config": [
      Obj({
        "installdir": [
          Str("Valheim dedicated server"),
        ],
        "launch": [
          Obj({
            "0": [
              Obj({
                "config": [
                  Obj({
                    "oslist": [
                      Str("linux"),
                    ],
                  }),
                ],
                "executable": [
                  Str("start_server_xterm.sh"),
                ],
                "type": [
                  Str("server"),
                ],
              }),
            ],
            "1": [
              Obj({
                "config": [
                  Obj({
                    "oslist": [
                      Str("windows"),
                    ],
                  }),
                ],
                "executable": [
                  Str("start_headless_server.bat"),
                ],
                "type": [
                  Str("server"),
                ],
              }),
            ],
          }),
        ],
      }),
    ],
    "depots": [
      Obj({
        "1004": [
          Obj({
            "config": [
              Obj({
                "oslist": [
                  Str("windows"),
                ],
              }),
            ],
            "depotfromapp": [
              Str("1007"),
            ],
            "manifests": [
              Obj({
                "public": [
                  Str("6473168357831043306"),
                ],
              }),
            ],
            "maxsize": [
              Str("39546856"),
            ],
            "name": [
              Str("Steamworks SDK Redist (WIN32)"),
            ],
          }),
        ],
        "1005": [
          Obj({
            "config": [
              Obj({
                "oslist": [
                  Str("macos"),
                ],
              }),
            ],
            "depotfromapp": [
              Str("1007"),
            ],
            "manifests": [
              Obj({
                "public": [
                  Str("2135359612286175146"),
                ],
              }),
            ],
            "name": [
              Str("Steamworks SDK Redist (OSX32)"),
            ],
          }),
        ],
        "1006": [
          Obj({
            "config": [
              Obj({
                "oslist": [
                  Str("linux"),
                ],
              }),
            ],
            "depotfromapp": [
              Str("1007"),
            ],
            "manifests": [
              Obj({
                "public": [
                  Str("6688153055340488873"),
                ],
              }),
            ],
            "maxsize": [
              Str("59862244"),
            ],
            "name": [
              Str("Steamworks SDK Redist (LINUX32)"),
            ],
          }),
        ],
        "896661": [
          Obj({
            "config": [
              Obj({
                "oslist": [
                  Str("linux"),
                ],
              }),
            ],
            "encryptedmanifests": [
              Obj({
                "experimental": [
                  Obj({
                    "encrypted_gid_2": [
                      Str("BEDF872D73873D16C025EF87E27C2BDB"),
                    ],
                    "encrypted_size_2": [
                      Str("2559486959C6E5DCEA5C71ED32BA9080"),
                    ],
                  }),
                ],
              }),
            ],
            "manifests": [
              Obj({
                "public": [
                  Str("521795651741005384"),
                ],
              }),
            ],
            "maxsize": [
              Str("985409357"),
            ],
            "name": [
              Str("Valheim dedicated server Linux"),
            ],
          }),
        ],
        "896662": [
          Obj({
            "config": [
              Obj({
                "oslist": [
                  Str("windows"),
                ],
              }),
            ],
            "encryptedmanifests": [
              Obj({
                "experimental": [
                  Obj({
                    "encrypted_gid_2": [
                      Str("9FD2B7B42FACB1D1FC439DD83ED2BED9"),
                    ],
                    "encrypted_size_2": [
                      Str("B2D602E667364DEDCB7C3D6EE9AA7374"),
                    ],
                  }),
                ],
              }),
            ],
            "manifests": [
              Obj({
                "public": [
                  Str("5449924312569304795"),
                ],
              }),
            ],
            "maxsize": [
              Str("963189471"),
            ],
            "name": [
              Str("Valheim dedicated server Windows"),
            ],
          }),
        ],
        "branches": [
          Obj({
            "experimental": [
              Obj({
                "buildid": [
                  Str("6263839"),
                ],
                "description": [
                  Str("Experimental version of Valheim"),
                ],
                "pwdrequired": [
                  Str("1"),
                ],
                "timeupdated": [
                  Str("1613728251"),
                ],
              }),
            ],
            "public": [
              Obj({
                "buildid": [
                  Str("6246034"),
                ],
                "timeupdated": [
                  Str("1613558776"),
                ],
              }),
            ],
            "unstable": [
              Obj({
                "buildid": [
                  Str("6246034"),
                ],
                "description": [
                  Str("Unstable test version of valheim"),
                ],
                "pwdrequired": [
                  Str("1"),
                ],
                "timeupdated": [
                  Str("1613469743"),
                ],
              }),
            ],
          }),
        ],
      }),
    ],
  }),
)
